hazelcast/proxy/cp/semaphore.py

```diff
@@ -0,0 +1,532 @@
+                acquired
+
+        Raises:
+            AssertionError: If the ``permits`` is not positive.
+        """
+        raise NotImplementedError("try_acquire")
+
+    def _do_change_permits(self, permits):
+        raise NotImplementedError("_do_change_permits")
+
+
+class SessionAwareSemaphore(Semaphore, SessionAwareCPProxy):
+    def acquire(self, permits=1):
+        check_true(permits > 0, "Permits must be positive")
+        current_thread_id = thread_id()
+        invocation_uuid = uuid.uuid4()
+        return self._do_acquire(current_thread_id, invocation_uuid, permits)
+
+    def drain_permits(self):
+        current_thread_id = thread_id()
+        invocation_uuid = uuid.uuid4()
+        return self._do_drain(current_thread_id, invocation_uuid)
+
+    def release(self, permits=1):
+        check_true(permits > 0, "Permits must be positive")
+        session_id = self._get_session_id()
+        if session_id == _NO_SESSION_ID:
+            return ImmediateExceptionFuture(self._new_illegal_state_error())
+
+        current_thread_id = thread_id()
+        invocation_uuid = uuid.uuid4()
+
+        def check_response(response):
+            try:
+                response.result()
+            except SessionExpiredError as e:
+                self._invalidate_session(session_id)
+                raise self._new_illegal_state_error(e)
+            finally:
+                self._release_session(session_id, permits)
+
+        return self._request_release(session_id, current_thread_id, invocation_uuid, permits).continue_with(
+            check_response)
+
+    def try_acquire(self, permits=1, timeout=0):
+        check_true(permits > 0, "Permits must be positive")
+        timeout = max(0, timeout)
+        current_thread_id = thread_id()
+        invocation_uuid = uuid.uuid4()
+        return self._do_try_acquire(current_thread_id, invocation_uuid, permits, timeout)
+
+    def _do_acquire(self, current_thread_id, invocation_uuid, permits):
+        def do_acquire_once(session_id):
+            session_id = session_id.result()
+
+            def check_response(response):
+                try:
+                    response.result()
+                except SessionExpiredError:
+                    self._invalidate_session(session_id)
+                    return self._do_acquire(current_thread_id, invocation_uuid, permits)
+                except WaitKeyCancelledError:
+                    self._release_session(session_id, permits)
+                    error = IllegalStateError("Semaphore(\"%s\") not acquired because the acquire call on the CP "
+                                              "group is cancelled, possibly because of another indeterminate call "
+                                              "from the same thread." % self._object_name)
+                    raise error
+                except Exception as e:
+                    self._release_session(session_id)
```

Comment:
I don't see this release session logic in Java client's `SessionAwareSemaphoreProxy`. Is it a leftover? If so, it makes sense to submit a PR to fix it in Java client and apply the change to other clients as well.

The same question applies to other methods with `except Exception as e` clause, like `_do_drain`.

Reply:
These are not leftovers, I forgot to send a PR to the Java side. I will do that and share the links within the team channel so that it can be applied to other clients. I also found that I was removing only 1 permit in case of an unspecified error on `_do_acquire`. Also fixed that and updated the test